### PR TITLE
C#: Extract enum underlying type from IL

### PR DESCRIPTION
**Pull request by tamasvajk**
Previously we were reporting `int` on CIL extracted enum underlying types. This PR changes this by extracting the correct underlying type from `TypeDefinitionHandle`s. Additionally, `System.Enum` is removed from the `Enum` QL class.

~~This PR adds commits on top of #4758.~~

~~[C# differences](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/677/)~~ [C# differences](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/678/)
[C# stats](https://jenkins.internal.semmle.com/job/Stats/job/CSharp/14/)

### CODE REVIEW

**Comment by tamasvajk** [comment]
not to self: missing DB upgrade folder, new stats, tests

**Comment by tamasvajk** [comment]
@hvitved The new enum test is failing. It looks like we're extracting different versions of CoreLib in CI, than what I have locally. Do you have any idea why that can be?

**Comment by tamasvajk** [comment]
[Stats job](https://jenkins.internal.semmle.com/job/Stats/job/CSharp/12/)

**Comment by tamasvajk** [comment]
> [Stats job](https://jenkins.internal.semmle.com/job/Stats/job/CSharp/12/)

[New stats job](https://jenkins.internal.semmle.com/job/Stats/job/CSharp/14/)

File: csharp/extractor/Semmle.Extraction.CIL/Entities/TypeDefinitionType.cs [hunk 1]
```diff
@@ -226,6 +235,59 @@ public override IEnumerable<IExtractionProduct> Contents
...
+                !namespaceName.IsNil &&
+                Cx.MdReader.StringComparer.Equals(namespaceName, "System");
+        }
+
+        internal PrimitiveTypeCode? GetUnderlyingEnumType()
+        {
+            foreach (var handle in td.GetFields())
+            {
+                var field = Cx.MdReader.GetFieldDefinition(handle);
+                if ((field.Attributes & FieldAttributes.Static) != 0)
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CIL/Entities/TypeDefinitionType.cs:274]:
```suggestion
                if (field.Attributes.HasFlag(FieldAttributes.Static))
```

File: csharp/ql/src/semmle/code/cil/Types.qll [hunk 2]
```diff
@@ -47,8 +47,12 @@ class ValueOrRefType extends DotNet::ValueOrRefType, Type, @cil_valueorreftype {
 class Enum extends ValueOrRefType {
   Enum() { this.isEnum() }
 
-  // Note that we don't actually use the proper internal representation yet.
-  override IntType getUnderlyingType() { any() }
+  override IntegralType getUnderlyingType() {
+    cil_enum_underlying_type(this, result)
+    or
+    not exists(IntegralType underlying | cil_enum_underlying_type(this, underlying)) and
+    result = any(IntType i)
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/cil/Types.qll:54]:
```suggestion
    not cil_enum_underlying_type(this, _) and
    result instanceof IntType
```

**Review by hvitved — CHANGES_REQUESTED** [review]
Minor comments, otherwise LGTM.

**Review by hvitved — DISMISSED** [review]

**Comment by hvitved** [comment]
Let's run this past CSharp-Differences before merging.

**Comment by tamasvajk** [comment]
I've rebased this PR, and pushed two extra commits (DB upgrade folder and change notes)

**Comment by tamasvajk** [comment]
There are no changes in the diff job.

**Review by hvitved — APPROVED** [review]

**Pull request merged**